### PR TITLE
Add support for Partitioning and Sample pages on mysql reflected options

### LINKED ISSUE #4038: add "PARTITION BY" and "STATS SAMPLE PAGES" to mysql reflected options
**Migrated issue, originally created by chebee7i ([@chebee7i](https://github.com/chebee7i))**

Some keyword kwargs like 'mysql_partition_by' and 'mysql_stats_sample_pages' are not reflected and stored in kwargs.

```
import sqlalchemy as sa
m = sa.MetaData()

t = sa.Table('ti', m,
    sa.Column('id', sa.Integer),
    sa.Column('amount', sa.DECIMAL(7, 2)),
    sa.Column('tr_date', sa.Date),
    mysql_engine='InnoDB',
    mysql_partition_by='HASH( MONTH(tr_date) )',
    mysql_partitions='6'
)

engine = ...
t.create(engine)
m2 = sa.Metadata(bind=engine, reflect=True)
assert 'mysql_partition_by' in m2.tables['ti'].kwargs  # AssertionError
```

### PR DESCRIPTION
**Pull request by RamonWill**
A user requested that keyword arguments such as mysql_partition_by and mysql_stats_sample_pages are reflected and stored in the kwargs when reflected. This fix reflects these as well as other keywords that relate to partitioning.

### Description
The proposed PR has additional private methods that will be able to perform regex on partition related lines, parse these lines and finally add these lines to the table options.
In our conversation beforehand I mentioned creating a partition_options attribute but i think doing so might require changes than i anticipated in the reflection process. So as a result i thought it would be easier to just use the existing table_options attribute. 

Note: Partition and Subpartition keywords are stored in the dialect options.

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [X ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**
Fixes: #4038

### CODE REVIEW

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **RamonWill** to try to get revision f8852cabe15c9a91de85d27980988051f7a1306d of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change f8852cabe15c9a91de85d27980988051f7a1306d: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2185

**Comment by sqla-tester** [comment]
**Ramon Williams** (RamonWill) wrote:

Hi Mike,

I have a two questions on the CI tests that are failing in Jenkins:

1. Locally i am testing on mysql version 8 and my tests all pass since i based them on this version. However, the Jenkins tests for versions 5.6/5.7 have failed mainly due to Letter case differences in these versions and version 8. I saw the bug you raised on a similar scenario here https://github.com/sqlalchemy/sqlalchemy/issues/4344. Should i overcome this by basing my tests  on version 5.6/5.7? 
Examples:
partition_by on versions 5.6/5.7 equals'RANGE (YEAR(c2))'
partition_by on version 8 equals 'RANGE (year(`c2`))'


2. For the mariadb tests that are failing would i have to follow the test_reflection_with_table_options() test and use "if testing.against("mariadb")'?

Thank you for your time,

Ramon

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2185

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

> Patch Set 1:
> 
> Hi Mike,
> 
> I have a two questions on the CI tests that are failing in Jenkins:
> 
> 1. Locally i am testing on mysql version 8 and my tests all pass since i based them on this version. However, the Jenkins tests for versions 5.6/5.7 have failed mainly due to Letter case differences in these versions and version 8. I saw the bug you raised on a similar scenario here https://github.com/sqlalchemy/sqlalchemy/issues/4344. Should i overcome this by basing my tests  on version 5.6/5.7? 
> Examples:
> partition_by on versions 5.6/5.7 equals'RANGE (YEAR(c2))'
> partition_by on version 8 equals 'RANGE (year(`c2`))'

Hi Ramon -

for the string comparison, I would take the approach of normalizing the string value for the comparison:

    def _normstr(value):
        return value.replace("`", "").lower()

    eq_(_normstr(kw['mysql_partition_by']), "range (year(c2))")


> 
> 
> 2. For the mariadb tests that are failing would i have to follow the test_reflection_with_table_options() test and use "if testing.against("mariadb")'?

What's happening there is that mariadb tests now run in a new "mariadb only mode" which is documented at https://docs.sqlalchemy.org/en/14/dialects/mysql.html#mariadb-only-mode .  this is giving us a little bit of room for the ongoing divergence of MariaDB from MySQL, while still not breaking out a whole new dialect for MariaDB since these two DBs are still very similar from an external point of view and it remains to be seen if they are going to actually diverge more.

in order to test "mariadb-only-mode" the test suite forces it on for any URL that points to a mariadb database.  so this adds a bit to the testing burden.   So the test you refer to test_reflection_with_table_options() is basically the form we have for now; you need to put the options in a second time, once with the "mysql_XYZ" prefix and a second time with "mariadb_XYZ" and then build out those conditionals.   this form is not great, but let's start with that just to get the tests passing here, and then let's see if we can come up with a more succinct pattern we can apply to both test_reflection_with_table_options() as well as the new tests here.   I can come up with something for that if you'd like.


> 
> Thank you for your time,
> 
> Ramon

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2185

**Comment by sqla-tester** [comment]
**Ramon Williams** (RamonWill) wrote:

Workflow+1

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2185

File: test/dialect/mysql/test_reflection.py [hunk 1]
```diff
@@ -375,13 +377,147 @@ def test_reflection_with_table_options(self):
...
             # assert reflected.kwargs['mysql_auto_increment'] == '5'
 
+    def test_reflection_with_partition_options(self):
+        kwargs = dict(
+            mysql_engine="InnoDB",
+            mysql_default_charset="utf8",
+            mysql_partition_by="HASH(MONTH(c2))",
+            mysql_partitions="6",
+        )
+
```
Comment (sqla-tester) [inline test/dialect/mysql/test_reflection.py:395]: **mike bayer** (zzzeek) wrote:

send in the connection fixture here and additionally @provide_metadata

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2185

**Review by sqla-tester — COMMENTED** [review]
**mike bayer** (zzzeek) wrote:

(9 comments)

looks good, I wonder if we can apply more modern patterns to the tests.  thanks!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2185

* test/dialect/mysql/test_reflection.py (line 532): ... and here...
* test/dialect/mysql/test_reflection.py (line 575): maybe the same way we have _norm_str(), we could have a method here _norm_reflected_table() ?  returns a dictionary with "partition_by" etc. inside of it so we don't have to have these big conditionals.  just a thought.

File: test/dialect/mysql/test_reflection.py [hunk 2]
```diff
@@ -375,13 +377,147 @@ def test_reflection_with_table_options(self):
...
+        def_table.create()
+        try:
+            reflected = Table("mysql_def", MetaData(testing.db), autoload=True)
+        finally:
+            def_table.drop()
+
+        assert def_table.kwargs["mysql_partition_by"] == "HASH(MONTH(c2))"
+        assert def_table.kwargs["mysql_partitions"] == "6"
+
+        assert reflected.kwargs["mysql_partition_by"] == "HASH (month(`c2`))"
```
Comment (sqla-tester) [inline test/dialect/mysql/test_reflection.py:413]: **mike bayer** (zzzeek) wrote:

self.metadata, due to @provide_metadata fixture

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2185

File: test/dialect/mysql/test_reflection.py [hunk 3]
```diff
@@ -375,13 +377,147 @@ def test_reflection_with_table_options(self):
...
+        assert reflected.kwargs["mysql_partitions"] == "6"
+
+    def test_reflection_with_subpartition_options(self):
+
+        subpartititon_text = """HASH (TO_DAYS (c2))
+                                SUBPARTITIONS 2(
```
Comment (sqla-tester) [inline test/dialect/mysql/test_reflection.py:419]: **mike bayer** (zzzeek) wrote:

self.metadata.create_all(connection)

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2185

File: test/dialect/mysql/test_reflection.py [hunk 4]
```diff
@@ -375,13 +377,147 @@ def test_reflection_with_table_options(self):
...
+                                 PARTITION p0 VALUES LESS THAN (1990),
+                                 PARTITION p1 VALUES LESS THAN (2000),
```
Comment (sqla-tester) [inline test/dialect/mysql/test_reflection.py:421]: **mike bayer** (zzzeek) wrote:

reflected = Table("mysql_def", MetaData(), autoload_with=connection)

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2185

File: test/dialect/mysql/test_reflection.py [hunk 5]
```diff
@@ -375,13 +377,147 @@ def test_reflection_with_table_options(self):
...
+                                 PARTITION p2 VALUES LESS THAN MAXVALUE
+                             );"""
```
Comment (sqla-tester) [inline test/dialect/mysql/test_reflection.py:423]: **mike bayer** (zzzeek) wrote:

take this out

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2185

File: test/dialect/mysql/test_reflection.py [hunk 6]
```diff
@@ -375,13 +377,147 @@ def test_reflection_with_table_options(self):
...
+        kwargs = dict(
+            mysql_engine="InnoDB",
+            mysql_default_charset="utf8",
+            mysql_partition_by="RANGE(YEAR(c2))",
+            mysql_subpartition_by=subpartititon_text,
+        )
```
Comment (sqla-tester) [inline test/dialect/mysql/test_reflection.py:429]: **mike bayer** (zzzeek) wrote:

use eq_(a, b) for all the "assert == " here

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2185

File: test/dialect/mysql/test_reflection.py [hunk 7]
```diff
@@ -375,13 +377,147 @@ def test_reflection_with_table_options(self):
...
+        def_table.create()
+        try:
+            reflected = Table("mysql_def", MetaData(testing.db), autoload=True)
+        finally:
+            def_table.drop()
+
+        assert def_table.kwargs["mysql_partition_by"] == "RANGE(YEAR(c2))"
+
+        ref_kwargs = reflected.kwargs
+        assert ref_kwargs["mysql_partition_by"] == "RANGE (year(`c2`))"
```
Comment (sqla-tester) [inline test/dialect/mysql/test_reflection.py:448]: **mike bayer** (zzzeek) wrote:

apply all the same changes as in test_reflection_with_partition_options to here

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2185

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

Code-Review-1

needs tests reworked

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2185

**Review by sqla-tester — COMMENTED** [review]
**Ramon Williams** (RamonWill) wrote:

Workflow+1

(9 comments)

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2185

* test/dialect/mysql/test_reflection.py (line 532): Done
* test/dialect/mysql/test_reflection.py (line 575): Good idea

File: test/dialect/mysql/test_reflection.py (hunk 1 above)
Comment (sqla-tester) [reply test/dialect/mysql/test_reflection.py:395]: **Ramon Williams** (RamonWill) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2185

File: test/dialect/mysql/test_reflection.py (hunk 2 above)
Comment (sqla-tester) [reply test/dialect/mysql/test_reflection.py:413]: **Ramon Williams** (RamonWill) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2185

File: test/dialect/mysql/test_reflection.py (hunk 3 above)
Comment (sqla-tester) [reply test/dialect/mysql/test_reflection.py:419]: **Ramon Williams** (RamonWill) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2185

File: test/dialect/mysql/test_reflection.py (hunk 4 above)
Comment (sqla-tester) [reply test/dialect/mysql/test_reflection.py:421]: **Ramon Williams** (RamonWill) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2185

File: test/dialect/mysql/test_reflection.py (hunk 5 above)
Comment (sqla-tester) [reply test/dialect/mysql/test_reflection.py:423]: **Ramon Williams** (RamonWill) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2185

File: test/dialect/mysql/test_reflection.py (hunk 6 above)
Comment (sqla-tester) [reply test/dialect/mysql/test_reflection.py:429]: **Ramon Williams** (RamonWill) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2185

File: test/dialect/mysql/test_reflection.py (hunk 7 above)
Comment (sqla-tester) [reply test/dialect/mysql/test_reflection.py:448]: **Ramon Williams** (RamonWill) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2185

**Comment by sqla-tester** [comment]
**Ramon Williams** (RamonWill) wrote:

Workflow+1

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2185

**Comment by sqla-tester** [comment]
**Ramon Williams** (RamonWill) wrote:

Workflow+1

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2185

**Comment by zzzeek** [comment]
it looks like i was looking for one more refinement to be made to the test suite, to get rid of the remaining conditionals with "is mariadb / is mysql".    It also needs a changelog file and a 1.4 backport.    so very small stuff.  Note that Ramon was working directly in gerrit so the content that's in the PR here is not the most recent.

**Comment by kkirsche** [comment]
Sounds good. I'll see if I can use the contributing guidelines on the website and in the repo to clone the Gerrit branch next week and see if this is something I can finalize.

**Comment by sqla-tester** [comment]
**Federico Caselli** (CaselIT) wrote:

changed branch to main created a new change. will continue at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4033

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2185

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2185 has been **abandoned**.  That means that at least for the moment I need to close this pull request. Sorry it didn't work out :(

**Comment by CaselIT** [comment]
> it looks like i was looking for one more refinement to be made to the test suite, to get rid of the remaining conditionals with "is mariadb / is mysql". It also needs a changelog file and a 1.4 backport. so very small stuff. Note that Ramon was working directly in gerrit so the content that's in the PR here is not the most recent.

I think we can skip the 1.4 backport

**Comment by sqla-tester** [comment]
**Federico Caselli** (CaselIT) wrote:

Can you take another look mike?

Looking at the PR is seems that most was done, I've just cleaned up the tests a bit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4033

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

great, looks like the round trip tests are very thorough, great!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4033

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4033 has been **merged**. Congratulations! :)

**Pull request closed without merging**